libs/pruviloj/Pruviloj/Core.idr

```diff
@@ -256,20 +256,23 @@ unproduct tm =
        try (unproduct (Var n1))
        try (unproduct (Var n2))
 
-||| If the goal is a tuple or a dependent pair, then split it into
-||| two goals and return the new goal names.
-||| It fails if the goal is neither a `Pair` nor a `DPair`.
-splitPair : Elab (TTName, TTName)
-splitPair =
-  let bug = [TextPart "Splitting created wrong number of holes"] in
-  case !goalType of
-    `(Pair ~A ~B)   => do [x, y] <- apply `(MkPair {A=~A} {B=~B}) [False, False]
-                                | _ => fail bug
-                          solve ; pure (x, y)
-    `(DPair ~A ~Pr) => do [x, y] <- apply `(MkDPair {a=~A} {P=~Pr}) [False, False]
-                              | _ => fail bug
-                          solve ; pure (x, y)
-    _ => fail [TextPart "Goal is not a pair"]
+||| Try to apply the constructors of the goal data type one by one,
+||| and apply the first one that works.
+||| Similar to `constructor` in Coq.
+construct : Elab (List TTName)
+construct =
+    do g <- goalType
+       case headName g of
+         Nothing =>
+           fail [RawPart g, TextPart "is not a type declared with the data keyword"]
+         Just h =>
+           choiceMap (\(n, xs, _) => apply (Var n) (map shouldUnify xs) <* solve)
+                     !(constructors <$> lookupDatatypeExact h)
+  where
+    shouldUnify : CtorArg -> Bool
```

Comment:
I like this heuristic. But it should be mentioned in the docstring that explicit constructor arguments will not be solved by unification.